### PR TITLE
NNS changes from C#

### PR DESCRIPTION
**Pull request by roman-khimov**
This updates and replaces #267. Most of the changes are already merged via other PRs, but:
 * #420 changes are duplicated in #267 and so are also present here
 * domain registration price/length dependency is omitted from this set because it doesn't fit our model at all

Otherwise they're OK to stay here for now (binaries/bindings not updated till we make decisions).

### CODE REVIEW

File: contracts/nns/contract.go [hunk 1]
```diff
@@ -629,7 +629,8 @@ func Resolve(name string, typ recordtype.Type) []string {
 	}
 
 	ctx := storage.GetReadOnlyContext()
-	return resolve(ctx, nil, name, typ, 2)
+	res := []string{}
```
Comment (cthulhu-rider) [inline contracts/nns/contract.go:652]: the compiler understands inline nil but not empty slice?

File: contracts/nns/contract.go [hunk 2]
```diff
@@ -461,16 +463,30 @@ func saveDomain(ctx storage.Context, name, email string, refresh, retry, expire,
 	putSoaRecord(ctx, name, email, refresh, retry, expire, ttl)
 }
 
-// Renew increases domain expiration date.
-func Renew(name string) int64 {
+// RenewDefault increases domain expiration date for 1 year.
```
Comment (cthulhu-rider) [inline contracts/nns/contract.go:466]: and returns what?

File: contracts/nns/contract.go [hunk 3]
```diff
@@ -461,16 +463,30 @@ func saveDomain(ctx storage.Context, name, email string, refresh, retry, expire,
...
+func RenewDefault(name string) int64 {
+	return Renew(name, 1)
+}
+
+// Renew increases domain expiration date up to the specified amount of years.
```
Comment (cthulhu-rider) [inline contracts/nns/contract.go:471]: same

File: contracts/nns/contract.go [hunk 4]
```diff
@@ -461,16 +463,30 @@ func saveDomain(ctx storage.Context, name, email string, refresh, retry, expire,
...
+func Renew(name string, years int) int64 {
+	if years < 1 || years > 10 {
+		panic("invalid renewal period value")
```
Comment (cthulhu-rider) [inline contracts/nns/contract.go:474]: limit not mentioned in docs

**Review by cthulhu-rider — COMMENTED** [review]
good overall

File: contracts/nns/contract.go (hunk 1 above)
Comment (carpawell) [reply contracts/nns/contract.go:652]: hm, yes, why not just `[]string{}` as an arg?

**Review by carpawell — DISMISSED** [review]

**Comment by roman-khimov** [comment]
Let's merge #420 then, some changes are present there and if we don't have any questions there it'd simplify things a bit.

File: contracts/nns/contract.go (hunk 1 above)
Comment (roman-khimov) [reply contracts/nns/contract.go:652]: Fixed in #420.

**Comment by carpawell** [comment]
@roman-khimov, done, can be rebased, also linter asks for attention.

**Review by carpawell — APPROVED** [review]

**Review by cthulhu-rider — APPROVED** [review]

**Pull request merged**